File: WikiCodeParser/Tags/WikiBookTag.cs
using WikiCodeParser.Nodes;
using WikiCodeParser.Models;

namespace WikiCodeParser.Tags
{
    public class WikiBookTag : Tag
    {
        public WikiBookTag()
        {
            Token = null;
            Element = "";
        }

        public override bool Matches(State state, string token)
        {
            var peekTag = state.Peek(6);
            var pt = state.PeekTo("]");
            return peekTag == "[book:" && pt != null && pt.Length > 6 && !pt.Contains("\n");
        }

        public override INode Parse(Parser parser, ParseData data, State state, string scope)
        {
            var index = state.Index;
            if (state.Next() != '[')
            {
                state.Seek(index, true);
                return null;
            }

            var str = state.ScanTo("]");

            if (state.Next() != ']')
            {
                state.Seek(index, true);
                return null;
            }

            var book = new WikiRevisionBook();

            var sections = str.Split('|');
            foreach (var section in sections)
            {
                var spl = section.Split(new[] {':'}, 2);
                var key = spl[0];
                var val = spl.Length > 1 ? spl[1] : "";
                switch (key) {
                    case "book":
                        book.BookName = val;
                        break;
                    case "chapter":
                        book.ChapterName = val;
                        break;
                    case "chapternumber":
                        if (int.TryParse(val, out var cn)) book.ChapterNumber = cn;
                        break;
                    case "pagenumber":
                        if (int.TryParse(val, out var pn)) book.PageNumber = pn;
                        break;
                }
            }

            state.SkipWhitespace();
            return new MetadataNode("WikiBook", book);
        }
    }
}

File: php/src/Tags/WikiBookTag.php
<?php

namespace LogicAndTrick\WikiCodeParser\Tags;

use LogicAndTrick\WikiCodeParser\Models\WikiRevisionBook;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\MetadataNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\State;
use LogicAndTrick\WikiCodeParser\TagParseContext;

class WikiBookTag extends Tag
{
    public function __construct()
    {
        parent::__construct();
        $this->token = null;
        $this->element = '';
    }

    public function Matches(State $state, ?string $token, TagParseContext $context): bool
    {
        $peekTag = $state->Peek(6);
        $pt = $state->PeekTo(']');
        return $peekTag == '[book:' && $pt != null && strlen($pt) > 6 && !str_contains($pt, "\n");
    }

    public function Parse(Parser $parser, ParseData $data, State $state, string $scope, TagParseContext $context): INode|null
    {
        $index = $state->index;
        if ($state->Next() != '[') {
            $state->Seek($index, true);
            return null;
        }

        $str = $state->ScanTo(']');

        if ($state->Next() != ']') {
            $state->Seek($index, true);
            return null;
        }

        $book = new WikiRevisionBook();

        $sections = explode('|', $str);
        foreach ($sections as $section) {
            $spl = explode(':', $section);
            $key = $spl[0];
            $val = count($spl) > 1 ? implode(':', array_slice($spl, 1)) : '';
            switch ($key) {
                case 'book':
                    $book->BookName = $val;
                    break;
                case 'chapter':
                    $book->ChapterName = $val;
                    break;
                case 'chapternumber':
                    $book->ChapterNumber = intval($val, 10) ?? null;
                    break;
                case 'pagenumber':
                    $book->PageNumber = intval($val, 10) ?? null;
                    break;
            }
        }

        $state->SkipWhitespace();
        return new MetadataNode('WikiBook', $book);
    }
}